### PR TITLE
Throw an exception with the asset key if a string asset load fails

### LINKED ISSUE #7715: CachingAssetBundle._fetchString gives poor error message
Ian asked me to report this bug. Loading the asset from the .flx file appears to be failing. It would be good if we were given the asset string whose load failed.

## Logs

```
[00206.770] 09972.10019> [ERROR:dart_error.cc(16)] Unhandled exception:
[00206.770] 09972.10019> NoSuchMethodError: The getter 'buffer' was called on null.
[00206.770] 09972.10019> Receiver: null
[00206.770] 09972.10019> Tried calling: buffer
[00206.770] 09972.10019> #0      Object._noSuchMethod (dart:core-patch/object_patch.dart:44)
[00206.770] 09972.10019> #1      Object.noSuchMethod (dart:core-patch/object_patch.dart:47)
[00206.770] 09972.10019> #2      CachingAssetBundle._fetchString.<_fetchString_async_body> (package:flutter/src/services/asset_bundle.dart:148)
[00206.770] 09972.10019> #3      _RootZone.runUnary (dart:async/zone.dart:1404)
[00206.770] 09972.10019> #4      _FutureListener.handleValue (dart:async/future_impl.dart:131)
[00206.770] 09972.10019> #5      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:637)
[00206.770] 09972.10019> #6      _Future._propagateToListeners (dart:async/future_impl.dart:667)
[00206.770] 09972.10019> #7      _Future._completeWithValue (dart:async/future_impl.dart:477)
[00206.770] 09972.10019> #8      _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:528)
[00206.770] 09972.10019> #9      _microtaskLoop (dart:async/schedule_microtask.dart:41)
[00206.770] 09972.10019> #10     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
[00206.835] 09972.10019>  EXCEPTION CAUGHT BY SERVICES 
[00206.839] 09972.10019> The following message was thrown resolving a single-frame image stream:
[00206.839] 09972.10019> Unable to read data
[00206.840] 09972.10019> 
[00206.840] 09972.10019> When the exception was thrown, this was the stack:
[00206.853] 09972.10019> #0      AssetBundleImageProvider.loadAsync.<loadAsync_async_body> (package:flutter/src/services/image_provider.dart:284)
[00206.853] 09972.10019> (elided 8 frames from package dart:async)
[00206.860] 09972.10019> 
[00206.860] 09972.10019> Image provider: AssetImage(bundle: null, name: "packages/armadillo/res/Background.jpg")
[00206.860] 09972.10019> Image key: AssetBundleImageKey(bundle: PlatformAssetBundle@686690892(), name:
[00206.860] 09972.10019> "packages/armadillo/res/Background.jpg", scale: 1.0)
[00206.860] 09972.10019> 
[00206.860] 09972.10019> Another exception was thrown: Unable to read data
[00206.860] 09972.10019> Another exception was thrown: Unable to read data
```

### PR DESCRIPTION
**Pull request by jason-simmons**
Fixes https://github.com/flutter/flutter/issues/7715

### CODE REVIEW

**Comment by jason-simmons** [comment]
@Hixie

**Comment by Hixie** [comment]
We should document in the dartdocs for load() and loadString() what the semantics are for the case where the asset is unavailable. Should they always throw? Always return null?

The fix in this PR only throws for the specific case of loadString() from an AssetBundle. It looks like the semantics for NetworkAssetBundle are to return null in the case of a successful load (!) for load(), and return null in the case of a failed load for loadString(); and for CachingAssetBundle, in case of failure load() returns null but loadString() throws. We should make this more consistent.

**Comment by jason-simmons** [comment]
Changed all load() implementations to throw on a missing asset

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 1]
```diff
@@ -51,6 +51,8 @@ import 'platform_messages.dart';
 ///  * [rootBundle]
 abstract class AssetBundle {
   /// Retrieve a binary resource from the asset bundle as a data stream.
+  ///
+  /// Throws an exception if the asset is not found.
```
Comment (Hixie) [inline packages/flutter/lib/src/services/asset_bundle.dart]: please add this to loadString's dartdocs as well

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 2]
```diff
@@ -97,15 +99,17 @@ class NetworkAssetBundle extends AssetBundle {
   @override
   Future<ByteData> load(String key) async {
     final http.Response response = await _httpClient.get(_urlFromKey(key));
-    if (response.statusCode == 200)
-      return null;
+    if (response.statusCode != 200)
+      throw new FlutterError('Unable to load asset: $key');
```
Comment (Hixie) [inline packages/flutter/lib/src/services/asset_bundle.dart]: This should probably include information from `response` as well (like the status code, maybe more? I guess we don't want to include the whole body, that might be absurdly verbose).

See <https://docs.flutter.io/flutter/foundation/FlutterError/message.html> for some style notes about these error messages.

**Comment by Hixie** [comment]
This is great.

Please add tests that verify the behaviour for the things for which you changed the behaviour.

LGTM.

**Comment by jason-simmons** [comment]
Added a test for AssetBundle - PTAL

File: packages/flutter/lib/src/services/asset_bundle.dart [hunk 3]
```diff
@@ -97,15 +101,19 @@ class NetworkAssetBundle extends AssetBundle {
...
   }
 
   @override
   Future<String> loadString(String key, { bool cache: true }) async {
     final http.Response response = await _httpClient.get(_urlFromKey(key));
-    return response.statusCode == 200 ? response.body : null;
+    if (response.statusCode != 200)
+      throw new FlutterError(
+          'Unable to load asset: $key\n'
+          'HTTP status code: ${response.statusCode}');
```
Comment (Hixie) [inline packages/flutter/lib/src/services/asset_bundle.dart]: nit: braces around multiline statements, and if you put a newline after the (, put a newline before the ).

File: packages/flutter/test/services/asset_bundle_test.dart [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    expect(bundle.loadCallCount['one'], 1);
+
+    FlutterError loadException;
+    try {
+      await bundle.loadString('foo');
+    } catch (e) {
+      loadException = e;
+    }
+    expect(loadException, isNotNull);
```
Comment (Hixie) [inline packages/flutter/test/services/asset_bundle_test.dart]: isFlutterError

**Comment by Hixie** [comment]
LGTM. Bonus points if you can also add a test for verifying the change to the network asset bundle... we really wouldn't want to regress the ==/!= 200 thing...

**Pull request merged**